File: subprojects/platform-play/src/testFixtures/resources/org/gradle/play/integtest/fixtures/app/advancedplayapp.old/app/models/DataType.java
/*
 * Copyright 2014 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package models;

public class DataType {
    private final String foo;
    private final int bar;

    public DataType(String foo, int bar) {
        this.foo = foo;
        this.bar = bar;
    }

    @Override
    public String toString() {
        return String.format("%s:%s", foo, bar);
    }
}

File: platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/language/DataType.kt
/*
 * Copyright 2024 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gradle.internal.declarativedsl.language

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable


interface DataType {
    sealed interface ConstantType<JvmType> : DataType
    @Serializable
    @SerialName("int")
    data object IntDataType : ConstantType<Int> {
        override fun toString(): String = "Int"
    }
    @Serializable
    @SerialName("long")
    data object LongDataType : ConstantType<Long> {
        override fun toString(): String = "Long"
    }
    @Serializable
    @SerialName("string")
    data object StringDataType : ConstantType<String> {
        override fun toString(): String = "String"
    }
    @Serializable
    @SerialName("boolean")
    data object BooleanDataType : ConstantType<Boolean> {
        override fun toString(): String = "Boolean"
    }

    // TODO: implement nulls?
    @Serializable
    @SerialName("null")
    data object NullType : DataType

    @Serializable
    @SerialName("unit")
    data object UnitType : DataType

    // TODO: `Any` type?
    // TODO: Support subtyping of some sort in the schema rather than via reflection?
}
